src/core/service.c

```diff
@@ -2140,6 +2141,83 @@ _pure_ static bool service_can_reload(Unit *u) {
         return !!s->exec_command[SERVICE_EXEC_RELOAD];
 }
 
+static unsigned service_exec_command_index(Unit *u, ServiceExecCommand id, bool control) {
+        Service *s = SERVICE(u);
+        unsigned index = 0;
+        ExecCommand *current, *first, *c;
+
+        assert(s);
+
+        first = s->exec_command[id];
+        current = control ? s->control_command : s->main_command;
+
+        /* Figure out where we are in the list by walking back to the beginning */
+        for (c = current; c != first; c = c->command_prev)
+                index++;
+
+        return index;
+}
+
+static int service_serialize_exec_command(Unit *u, FILE *f, bool control) {
+        Service *s = SERVICE(u);
+        ServiceExecCommand id;
+        ExecCommand *c;
+        int index;
+        const char *type;
+        char **arg;
+        _cleanup_strv_free_ char **escaped_args = NULL;
+        _cleanup_free_ char *args = NULL, *p = NULL;
+        size_t allocated = 0, length = 0;
+
+        assert(s);
+        assert(f);
+
+        if (control) {
+                type = "control";
+                c = s->control_command;
+                id = s->control_command_id;
+        } else {
+                type = "main";
+                c = s->main_command;
+                id = SERVICE_EXEC_START;
```

Comment:
similar as above, instead of taking a "control" bool, I'd just take these three values as function parameters, so that the caller picks them the right way